# Demonstration using the Breast cancer Wisconsin data

The Breast cancer Wisconsin dataset is a dataset that describe the characteristics of the cell nuclei for 569 breast lumps, and includes whether they are malignant or benign. Published by Clinical Science Center, University of Wisconsin.

Source: https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29

**This Jupyter notebook will put to the test the artificial neuron that was written from scratch without dependence on any ML libraries.**

In [1]:
import artificial_neuron as an
import numpy as np
import pandas as pd

from sklearn import model_selection, preprocessing

In [2]:
data=pd.read_csv("../datasets/breast-cancer-wisconsin-data/data.csv", delimiter=",")


In [3]:
display(data.head())
display(data.describe())

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

Convert the target ("daignosis" column) from nominal strings ("M" and "B") to numbers (0 and 1).

In [4]:
data["diagnosis"]=preprocessing.LabelEncoder().fit(["M","B"]).transform(data["diagnosis"]) # M is 0, B is 1.

data.drop(columns="id",inplace=True)

Convert the pandas Dataframe to numpy array.

In [5]:
data=data.to_numpy()

Split into train and test sets, and also seperate into features (x) and target (y). Also ensure the shape of the input and output tensors are as expected. For features: num_features x batch_size. For target: 1 x batch_size.

In [6]:
data_train, data_test = model_selection.train_test_split(data, train_size=0.8, random_state=5)

In [7]:
print(data_train.shape)
print(data_test.shape)

(455, 31)
(114, 31)


In [8]:
data_train=data_train.T
data_test=data_test.T

print(data_train.shape)
print(data_test.shape)

(31, 455)
(31, 114)


In [9]:
x_train=data_train[1:, :]
y_train=data_train[0, :].reshape(1,-1)

x_test=data_test[1:, :]
y_test=data_test[0, :].reshape(1,-1)

# Training and evaluation

Just start with some randomly chosen hyperparameters:

* number of iterations: 10

* learning rate (step size): 0.00007

* batch size: 256

In [10]:
neuron=an.Neuron(X=x_train, Y=y_train)
neuron.train(num_iterations=10, learning_rate=7e-5, batch_size=256)

print("training accuracy: ", neuron.evaluate(X=x_train, Y=y_train, metric="accuracy"))
print("test accuracy:     ", neuron.evaluate(X=x_test, Y=y_test, metric="accuracy"), end="\n\n")

Training begins...
Training Complete!
Caution: All the activations are null values.
training accuracy:  None
Caution: All the activations are null values.
test accuracy:      None



C:\Users\princ\OneDrive\Project Intelligent Machine\Writeups\Multilayer Perceptron\Deep_Learning_Without_ML_Libraries\one_neuron\artificial_neuron.py:49: RuntimeWarning: invalid value encountered in true_divide
  self.dJdA = -(1/m) * ((self.Y_batch / self.a) - ((1 - self.Y_batch) / (1 - self.a)))


Let's check the activations

In [11]:
print(neuron.a)

[[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan]]


We likely ran into an overflow or underflow somewhere which led to the NaNs. We can try to avoid this by tweaking our hyperparameters.

In [17]:
vaious_iterations=[50, 100, 500, 1000]
for n in vaious_iterations:
    neuron=an.Neuron(X=x_train, Y=y_train)
    neuron.train(num_iterations=n, learning_rate=7e-6, batch_size=32)

    print("training accuracy: ", neuron.evaluate(X=x_train, Y=y_train, metric="accuracy"))
    print("test accuracy:     ", neuron.evaluate(X=x_test, Y=y_test, metric="accuracy"), end="\n\n")

Training begins...
Training Complete!
training accuracy:  0.36043956043956044
test accuracy:      0.42105263157894735

Training begins...
Training Complete!
training accuracy:  0.8263736263736263
test accuracy:      0.8157894736842105

Training begins...
Training Complete!
training accuracy:  0.9120879120879121
test accuracy:      0.9122807017543859

Training begins...
Training Complete!
training accuracy:  0.9120879120879121
test accuracy:      0.8771929824561403



We can start seeing the training accuracy outpace the test accuracy by the time we are reaching 1000 iterations, so we stop there as that is possibly a sign that we are begining to overfit (a little overfitting is perfectly fine).

**We have 91% test accuracy after 500 iterations.**

Let's play around with the other hyperparameters to see if we can get something better.

In [18]:
learning_rates=[7e-7,7e-6,7e-5,7e-4]
for lr in learning_rates:    
    neuron=an.Neuron(X=x_train, Y=y_train)
    neuron.train(num_iterations=700, learning_rate=lr, batch_size=32)

    print("training accuracy: ", neuron.evaluate(X=x_train, Y=y_train, metric="accuracy"))
    print("test accuracy:     ", neuron.evaluate(X=x_test, Y=y_test, metric="accuracy"), end="\n\n")

Training begins...
Training Complete!
training accuracy:  0.8021978021978022
test accuracy:      0.8157894736842105

Training begins...
Training Complete!
training accuracy:  0.9120879120879121
test accuracy:      0.8859649122807017

Training begins...
Training Complete!
Caution: All the activations are null values.
training accuracy:  None
Caution: All the activations are null values.
test accuracy:      None

Training begins...
Training Complete!
Caution: All the activations are null values.
training accuracy:  None
Caution: All the activations are null values.
test accuracy:      None



We overflowed/underflowed when our learning rate got relatively too high.

Let's try out different batch sizes. 

In [25]:
# truncate the printout for np array.
np.set_printoptions(threshold=6)

batch_sizes=[8, 32, 64, 256]
for b in batch_sizes:
    neuron=an.Neuron(X=x_train, Y=y_train)
    neuron.train(num_iterations=500, learning_rate=7e-6, batch_size=b)
    
    print("training accuracy: ", neuron.evaluate(X=x_train, Y=y_train, metric="accuracy"))
    print("test accuracy:     ", neuron.evaluate(X=x_test, Y=y_test, metric="accuracy"))
    print(neuron.a, end="\n\n")

# resetting to the default.
np.set_printoptions(threshold=None)

Training begins...
Training Complete!
training accuracy:  0.9164835164835164
test accuracy:      0.9122807017543859
[[0.99337728 0.23600475 0.21366923 ... 0.88089114 0.57491837 0.75790684]]

Training begins...
Training Complete!
training accuracy:  0.9120879120879121
test accuracy:      0.9122807017543859
[[0.98380587 0.24402546 0.22197814 ... 0.82355827 0.53042739 0.69759733]]

Training begins...
Training Complete!
training accuracy:  0.9164835164835164
test accuracy:      0.8947368421052632
[[0.97663485 0.22171196 0.19468137 ... 0.76620896 0.49236971 0.62952985]]

Training begins...
Training Complete!
training accuracy:  0.9164835164835164
test accuracy:      0.8859649122807017
[[0.97083241 0.21702171 0.18807814 ... 0.73339687 0.47702715 0.59767561]]



At this learning are, we are getting good performance with less batch size, but this doesn't mean we will always do better with smaller batch sizes.

Let's shoot for the highest test accuracy possible.

In [65]:
neuron=an.Neuron(X=x_train, Y=y_train)
neuron.train(num_iterations=1300, learning_rate=9e-6, batch_size=256)

print("training accuracy: ", neuron.evaluate(X=x_train, Y=y_train, metric="accuracy"))
print("test accuracy:     ", neuron.evaluate(X=x_test, Y=y_test, metric="accuracy"), end="\n\n")

Training begins...
Training Complete!
training accuracy:  0.9186813186813186
test accuracy:      0.9210526315789473



**We achieved 92%**. Maybe we can get one or two more % if we keep trying.